sogs/model/room.py

```diff
@@ -979,6 +1042,207 @@ def attachments_size(self):
             "SELECT COUNT(*), COALESCE(SUM(size), 0) FROM files WHERE room = :r", r=self.id
         ).first()[0:2]
 
+    def get_reactions(
+        self,
+        message_ids: list[int],
+        user: Optional[User] = None,
+        *,
+        reactors: int = 0,
+        session_ids: bool = True,
+    ):
+        """
+        Returns reaction information for the given messages.  Returns a dict of message id ->
+        reaction info for any messages in the given list that have reactions.  The reaction info is
+        itself a dict where keys are reaction strings and values are a dict containing keys:
+
+        - `"count"` -- the total number of this reaction
+        - `"you"` -- boolean set to true if `user` is given and the user has applied this reaction;
+          omitted otherwise.
+        - `"reactors"` -- if the input parameter `reactors` is greater than 0 then this is a list of
+          the session ids (if `session_ids`) or user ids (`session_ids` false) of the first
+          `reactors` users that applied this reaction.  Omitted if `reactors` is 0.
+
+        This method does *not* check for read access of the given messages: this is designed for
+        internal use from message ids that have already been permission checked.
+        """
+
+        reacts = {}
+        you = (
+            """
+            EXISTS(
+            SELECT * FROM reactions WHERE message = r.message AND reaction = r.reaction
+                AND "user" = :uid)
+            """
+            if user is not None
+            else 'FALSE'
+        )
+        for msgid, react, count, you in query(
+            f"""
+            SELECT
+                message,
+                reaction,
+                COUNT(*) AS react_count,
+                {you} AS you
+            FROM reactions r
+            WHERE message IN :msgs GROUP BY message, reaction
+            """,
+            bind_expanding=('msgs',),
+            msgs=message_ids,
+            uid=None if user is None else user.id,
+        ):
+            reacts.setdefault(msgid, {})[react] = (
```

Comment:
can reacts be a `collections.defaultdict` instead?

Reply:
It could, but it wouldn't save anything: this is the only place we use `setdefault` on it.